google/cloud/storage/transfer_manager.py

```diff
@@ -0,0 +1,188 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import concurrent.futures
+
+import tempfile
+
+
+def upload_many(
+    file_blob_pairs,
+    skip_if_exists=False,
+    upload_kwargs=None,
+    max_workers=None,
+    deadline=None,
+    raise_exception=False,
+):
+    if upload_kwargs is None:
+        upload_kwargs = {}
+    if skip_if_exists:
+        upload_kwargs["if_not_generation_match"] = 0
+
+    with concurrent.futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
+        futures = []
+        for path_or_file, blob in file_blob_pairs:
+            method = (
+                blob.upload_from_filename
+                if isinstance(path_or_file, str)
+                else blob.upload_from_file
+            )
+            futures.append(executor.submit(method, path_or_file, **upload_kwargs))
+    results = []
+    concurrent.futures.wait(
+        futures, timeout=deadline, return_when=concurrent.futures.ALL_COMPLETED
+    )
+    for future in futures:
+        if not raise_exception:
+            exp = future.exception()
+            if exp:
+                results.append(exp)
+                continue
+        results.append(future.result())
```

Comment:
It should be impossible because the call `concurrent.futures.wait(futures, timeout=deadline, return_when=concurrent.futures.ALL_COMPLETED)` will ensure that every single future is complete before getting to that stage.